default/AI/ResearchAI.py

```diff
@@ -20,21 +32,402 @@ def get_research_index():
-            tech_name in TechsListsAI.unusable_techs())
-            
+def tech_time_sort_key(tech_name):
+    return research_reqs.get(tech_name, ([],0,0,0))[REQS_TIME_IDX]
 
 def generate_research_orders():
     """generate research orders"""
+
+    # initializing priority functions here within generate_research_orders() to avoid import race
+
+    DEFENSIVE = foAI.foAIstate.aggression <= fo.aggression.cautious
+
+    # keys are "PREFIX1", as in "DEF" or "SPY"
+    if not primary_prefix_priority_funcs:
+        primary_prefix_priority_funcs.update({
+            AIDependencies.DEFENSE_TECHS_PREFIX: partial(const_priority, 2.0) if DEFENSIVE else partial(if_enemies, 0.2)
+            })
+
+    # keys are "PREFIX1_PREFIX2", as in "SHP_WEAPON"
+    if not secondary_prefix_priority_funcs:
+        secondary_prefix_priority_funcs.update({
+            AIDependencies.WEAPON_PREFIX: partial(ship_usefulness, partial(if_enemies, 0.2), MIL_IDX)
+            })
+
+    # keys are lists of full tech names
+    if not misc_group_priority_funcs:
+        misc_group_priority_funcs.update({
+            tuple(AIDependencies.UNRESEARCHABLE_TECHS): partial(const_priority, -1.0),
+            tuple(AIDependencies.UNUSED_TECHS): priority_zero,
+            tuple(AIDependencies.THEORY_TECHS): priority_zero,
+            tuple(AIDependencies.PRODUCTION_BOOST_TECHS): partial(if_dict, ColonisationAI.empire_status, 'industrialists', 0.6, 1.5),
+            tuple(AIDependencies.RESEARCH_BOOST_TECHS): partial(if_tech_target, get_initial_research_target(), 2.1, 2.5),
+            tuple(AIDependencies.PRODUCTION_AND_RESEARCH_BOOST_TECHS): partial(const_priority, 2.5),
+            tuple(AIDependencies.POPULATION_BOOST_TECHS): get_population_boost_priority,
+            tuple(AIDependencies.SUPPLY_BOOST_TECHS): partial(if_tech_target, AIDependencies.SUPPLY_BOOST_TECHS[0], 1.0, 0.5),
+            tuple(AIDependencies.METER_CHANGE_BOOST_TECHS): partial(const_priority, 1.0),
+            tuple(AIDependencies.DETECTION_TECHS): partial(const_priority, 0.5),
+            tuple(AIDependencies.STEALTH_TECHS): get_stealth_priority,
+            tuple(AIDependencies.DAMAGE_CONTROL_TECHS): partial(if_enemies, 0.1, 0.5),
+            tuple(AIDependencies.HULL_TECHS): get_hull_priority,
+            tuple(AIDependencies.ARMOR_TECHS): partial(ship_usefulness, if_enemies, MIL_IDX),
+            tuple(AIDependencies.ENGINE_TECHS): partial(ship_usefulness, partial(if_dict, choices, 'engine', true_val=0.6), None),
+            tuple(AIDependencies.FUEL_TECHS): partial(ship_usefulness, partial(if_dict, choices, 'fuel'), None),
+            tuple(AIDependencies.SHIELD_TECHS): partial(ship_usefulness, if_enemies, MIL_IDX),
+            tuple(AIDependencies.TROOP_POD_TECHS): partial(ship_usefulness, partial(if_enemies, 0.1, 0.3), TROOP_IDX),
+            tuple(AIDependencies.COLONY_POD_TECHS): partial(ship_usefulness, partial(const_priority, 0.5), COLONY_IDX),
```

Comment:
Should we not define the tech groups in AIDependencies directly as a tuple?

it would make the dictionary definition slightly easier to read and the lines somewhat shorter (the syntax-coloring here is also slightly distracting).